### ISSUE TITLE
class-decorator errors but equivalent function call succeeds

### ISSUE THREAD

**Issue by randolf-scholz** [microsoft/pyright#8681]
In the example below, `pyright` complains that the decorator is untyped and claims that `PointA` lacks the `__dataclass_fields__` attribute. Yet, when the decorator is applied as a function to the equivalent class `PointB`, no such diagnostics are produced, so presumably either one must be incorrect. ([mypy playground](https://mypy-play.net/?mypy=latest&python=3.12&enable-incomplete-feature=NewGenericSyntax&flags=strict&gist=3ac012eb6d4006f76f396609bad677fe) with `--strict` shows no errors)

Code sample in [pyright playground](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAK4MYAxmADYA0UAwhQIYDOTAagyDQIIpwBQoSFAAmDGA1KMWAUyaZsufKPGTmTGgDEk0isL59VLKABExEqUwAURMCXIUAlAC4%2BUN1AD6H5ebUfg2rpMXk50FuwgANrCSKQwkUwwnFBaOsKRPHAAujmu7gBEhQAqABbSUAFpcmDAsGUiZoZMAHSF%2BfrC0rU%2BTd7S5CBiuJHGoaYqFlmWkkyh8AjSI1kOUAC0AHywiIvGWS7uUADEFbhQ0gAeDNgU0jQMwsJQDFCkAK6JQggg0jDwUF%2BoGCoDAAI2Y0keYBQdXKlSCzTybm%2BMFeIGhM30x3ewJhIn6uCGICecmenQGhL4AAEehY%2BuSSESjlBADLkUEGSCYsjZ0hweC4IDQrwg0hQMCK2xo315MAAqqLtsJ6GpjPjBgyqTS1AYLIQwICuPt3OdQoDEVA4CbRfpqY0LNq1LrAQAhQ1uY2YK0HC0emCYqDY9C4skEhnEp4VV4oOJIKF8Ah60VQAC8DQmfmDatw1gTMCdK2OgFByKAoMBnEDgIl8IA)

```python
from typing import Protocol, ClassVar, Any
from dataclasses import dataclass, Field

class Dataclass(Protocol):
    __dataclass_fields__: ClassVar[dict[str, Field[Any]]]
    """The fields of the dataclass."""

def dataclass_decorator[D: Dataclass](cls: type[D]) -> type[D]:
    # for example, add a custom pretty printing based on the fields.
    return cls

# using the decorator as a decorator
@dataclass_decorator  # ❌ raises reportArgumentType, reportUntypedClassDecorator
@dataclass
class PointA:
    x: int
    y: int

@dataclass
class PointB:
    x: int
    y: int

# using the decorator as a function
Point = dataclass_decorator(PointB) # ✅ no error 

```

Maybe related to https://github.com/microsoft/pyright/issues/4339? The typeshed source contains [a comment](https://github.com/python/typeshed/blob/8a7f09e3511f3a1d04281c60167b8dcc3b78938b/stdlib/_typeshed/__init__.pyi#L345-L350) that such a Dataclass Protocol might produce unexpected results in `pyright`.

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

You've created a circular reference here that cannot be resolved.

The issue is that dataclasses can (and do) contain fields that refer to the dataclass type itself. For example, `PointA` in your example, could contain a field defined as `a: "PointA"`. In this case, `"PointA"` refers to the decorated `PointA` — i.e. the version of the class that has all of its class decorators applied. For that reason, pyright processes class decorations on the class before processing any of the fields, whose types affect synthesized dataclass methods. That means `__dataclass_fields__` has not yet been synthesized at the point where your `dataclass_decorator` call is evaluated for `PointA`.

In the case of `PointB`, you have broken the circularity by applying the decorator independently. This is the approach I recommend if your dataclass class decorator depends on the presence of synthesized methods and fields.
